### DISCUSSION TITLE
Using Build Spec to install `yarn`

### DISCUSSION THREAD

**Discussion by truecolintracy** [CloudSnorkel/cdk-github-runners#247] [Q&A] [answered]
I'm a little confused as to how to properly install `yarn` on the docker image. I don't properly understand how to extend this construct to implement something like:

```
aws_codebuild.BuildSpec.fromObject({
    version: '0.2',
    phases: {
      install: {
        'runtime-versions': {
          nodejs: 18,
        },
        commands: [
          'echo "Installing dependencies"',
          'npm install -g yarn',
          'yarn',
        ],
      },
    },
  });
```

**Comment by truecolintracy** [comment]
Understanding now that is only for use with the actual runner and I need to install `yarn` on the actual docker instance. Still a bit lost on how to accomplish this, I'm thinking it's pretty simple but I'm just not seeing it. :|

**Comment by kichik** [comment, marked as answer]
You would have to configure your runner image builder to build images with your additional requirements. If you are building AMI or a container image for Windows, it's pretty straightforward (and copied from [README.md](https://github.com/CloudSnorkel/cdk-github-runners#customizing)).

```typescript
const myWindowsBuilder = new ContainerImageBuilder(this, 'Windows image builder', {
  architecture: Architecture.X86_64,
  os: Os.WINDOWS,
  runnerVersion: RunnerVersion.specific('2.291.0'),
  rebuildInterval: Duration.days(14),    
});
myWindowsBuilder.addComponent(new ImageBuilderComponent(this, 'Ninja Component',
  {
    displayName: 'Ninja',
    description: 'Download and install Ninja build system',
    platform: 'Windows',
    commands: [
      'Invoke-WebRequest -UseBasicParsing -Uri "https://github.com/ninja-build/ninja/releases/download/v1.11.1/ninja-win.zip" -OutFile ninja.zip',
      'Expand-Archive ninja.zip -DestinationPath C:\\actions',
      'del ninja.zip',
    ],
  }
));

const myProvider = new FargateRunnerProvider(this, 'fargate runner', {
  label: 'customized-windows-fargate',
  vpc: vpc,
  securityGroup: runnerSg,
  imageBuidler: myWindowsBuilder,
});

// create the runner infrastructure
new GitHubRunners(this, 'runners', {
  providers: [myProvider],
});
```

Linux would be almost identical (if you're building an AMI):

```typescript
const myBuilder = new AmiBuilder(this, 'AMI builder', {
  architecture: Architecture.X86_64,
  os: Os.LINUX,
});
myBuilder.addComponent(new ImageBuilderComponent(this, 'Yarn',
  {
    displayName: 'Yarn',
    description: 'Download and install Yarn',
    platform: 'Linux',
    commands: [
      'npm install -g yarn',
    ],
  }
));

const myProvider = new Ec2RunnerProvider(this, 'ec2 runner with yarn', {
  labels: ['ec2-with-yarn'],
  imageBuidler: myBuilder,
});

// create the runner infrastructure
new GitHubRunners(this, 'runners', {
  providers: [myProvider],
});
```

If you want to build a runner image for Linux with a provider that needs a Docker image, it's a bit more complicated. First figure out which Dockerfile you want to modify. Each provider has two constants pointing to the paths of Dockerfiles it uses by default. `CodeBuildRunnerProvider` has one for Linux x64 and one for ARM64.

https://github.com/CloudSnorkel/cdk-github-runners/blob/a62cecf8ac65e453a78638c6f170f3ab82f7d76b/src/providers/codebuild.ts#L123-L147

All of the Dockerfiles are under the same [folder](https://github.com/CloudSnorkel/cdk-github-runners/tree/e4ebc40016d22f31a3555b74c74797fdb6a99e1f/src/providers/docker-images). For example, if you're trying to create a CodeBuild runner provider on x64, it's [here](https://github.com/CloudSnorkel/cdk-github-runners/blob/main/src/providers/docker-images/codebuild/linux-x64/Dockerfile). 

Most Dockerfiles have some arguments you can override. Those are also documented in the constants pointing to them from the provider as you can see above. If you just want to install an extra package, there is an argument to override for that.

https://github.com/CloudSnorkel/cdk-github-runners/blob/e4ebc40016d22f31a3555b74c74797fdb6a99e1f/src/providers/docker-images/codebuild/linux-x64/Dockerfile#L8-L10

```typescript
const builder = new CodeBuildImageBuilder(stack, 'codebuild image builder', {
  dockerfilePath: CodeBuildRunnerProvider.LINUX_X64_DOCKERFILE_PATH,
  architecture: Architecture.X86_64,
});
builder.setBuildArg('EXTRA_PACKAGES', 'yarnpkg');

new GitHubRunners(this, 'runners', {
  providers: [
    new CodeBuildRunnerProvider(this, 'provider', {
      imageBuilder: builder,
    });
  ],
});
```

If you want to add arbitrary commands, you will have to completely override the Dockerfile. This can be useful in your case if the version of Yarn that comes with Ubuntu is not new enough. In this case, you would make a copy of the relevant Dockerfile and modify it to your liking. You would then point the builder to your Dockerfile directory instead of `CodeBuildRunnerProvider.LINUX_X64_DOCKERFILE_PATH`.

In the future I plan to make a generic component that can be added to both Docker image builders and AMI builders. This should make this process easier. There will be no more need to hunt down and copy any Dockerfile.

**Comment by kichik** [comment]
With version 0.9.0 you can now use the following on all runner providers:

```
const builder = CodeBuildRunnerProvider.imageBuilder(stack, 'codebuild image builder');
builder.addComponent(
  RunnerImageComponent.custom({ commands: ['apt install -y yarnpkg'] })
);

new GitHubRunners(this, 'runners', {
  providers: [
    new CodeBuildRunnerProvider(this, 'provider', {
      imageBuilder: builder,
    });
  ],
});
```
